lib/livebook/zta/teleport.ex

```diff
@@ -0,0 +1,92 @@
+defmodule Livebook.ZTA.Teleport do
+  use GenServer
+  require Logger
+
+  defstruct [:name, :req_options, :jwks]
+
+  @renew_afer 24 * 60 * 60 * 1000
+  @fields %{"sub" => :id, "username" => :username}
+  @assertion "teleport-jwt-assertion"
+  @well_known_jwks_path "/.well-known/jwks.json"
+
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts, name: opts[:name])
+  end
+
+  def authenticate(name, conn, _) do
+    token = Plug.Conn.get_req_header(conn, @assertion)
+
+    jwks = GenServer.call(name, :get_jwks, :infinity)
+
+    {conn, authenticate_user(token, jwks)}
+  end
+
+  @impl true
+  def init(opts) do
+    url =
+      opts[:identity][:key]
+      |> URI.parse()
+      |> URI.append_path(@well_known_jwks_path)
+      |> URI.to_string()
```

Comment:
Parse this on start_link if you can. If something goes wrong, it leads to less nested error messages. :)

Reply:
Agree! I moved it here because I thought this logic doesn't belong to `start_link` but ofc I've see errors like what you mentioned and nobody wants to read that 😬